proposals/gc/Overview.md

```diff
@@ -1 +1,716 @@
+* References are never null; nullable reference types are separate.
+
+
+### Structures
+
+*Structure* types define aggregates with heterogeneous fields that are _statically indexed_:
+```
+(type $time (struct (field i32) (field f64)))
+(type $point (struct (field $x f64) (field $y f64) (field $z f64)))
+```
+Such types can be used by forming *reference types*, which are a new type of value type:
+```
+(local $var (ref $point))
+```
+
+Fields are *accessed* with generic load/store instructions that take a reference to a structure:
+```
+(func $f (param $p (ref $point))
+  (store_field $point $y (get_local $p)
+    (load_field $point $x (get_local $p))
+  )
+)
+```
+All accesses are type-checked at validation time.
+
+Structures are *allocated* witha  `new` instruction that take initialization values for each field.
+The operator yields a reference to the respective type:
+```
+(func $g
+  (call $g (new_struct $point (i32.const 1) (i32.const 2) (i32.const 3)))
+)
+```
+Structures are garbage-collected.
+
+Structures can be compared for identity:
+```
+(same (new $point ...) (new $point ...))  ;; false
+```
+TODO: Could even allow heterogeneous equality (equality between operands of different type), but that might lead to some discontinueties or even prevent some potential optimizations?
+
+
+### Arrays
+
+*Array* types define aggregates with _homogeneous elements_ that are _dynamically indexed_:
+```
+(type $vector (array f64))
+(type $matrix (array (type $vector)))
+```
+Such types again can be used by forming reference types.
+For now, we assume that all array types have dynamic ([flexible](#flexible-aggregates)) size.
+
+Elements are accessed with generic load/store instructions that take a reference to an array:
+```
+(func $f (param $v (ref $vector))
+  (store_elem $vector (get_local $v) (i32.const 1)
+    (load_elem $vector (get_local $v) (i32.const 2))
+  )
+)
+```
+The element type of every access is checked at validation time.
+The index is checked at execution time.
+A trap occurs if the index is out of bounds.
+
+Arrays are *allocated* with a `new` instruction that takes a size and an initialization value as operands, yielding a reference:
+```
+(func $g
+  (call $g (new_array $vector (i32.const 0) (i64.const 3.14)))
+)
+```
+Arrays are garbage-collected.
+
+The *length* of an array, i.e., the number of elements, can be inquired via the `load_length` instruction:
+```
+(load_length $vector (get_local $v))
+```
+
+Like structures, arrays can be compared for identity.
+
+
+### Packed Fields
+
+Fields and elements can have a packed *storage type* `i8` or `i16`:
+```
+(type $s (struct (field $a i8) (field $b i16)))
+(type $buf (array i8))
+```
+The order of fields is not observable, so implementations are free to optimize types by reordering fields or adding gaps for alignment.
+
+Packed fields require special load/store instructions:
+```
+(load_field_packed_s $s $a (...))
```

Comment:
Wasm does not distinguish signed from unsigned integer types, so it is necessary here to specify sign extension behaviour, same as with existing instructions such as memory loads and stores. That also implies that at least the packed loads are different opcodes than non-packed loads. For symmetry and clarity it seems preferable to do the same for stores then.